From 2aef713a5f9e30e52f3e3631d260c850e4ac9fdf Mon Sep 17 00:00:00 2001
From: Aaron T. Myers <atm@cloudera.com>
Date: Tue, 14 May 2013 15:06:29 -0700
Subject: [PATCH 1340/1357] CLOUDERA-BUILD. Explicitly throw an UnsupportedOperationException if
 newer client is running against older server which does not support
 the isFileClosed feature.

Ref: CDH-11316
Author: Aaron T. Myers
Reason: Allow users of this API to easily detect if their client is
incompatible with the server.
---
 .../apache/hadoop/hdfs/DistributedFileSystem.java  |    3 +++
 .../ClientNamenodeProtocolTranslatorPB.java        |    8 +++++++-
 2 files changed, 10 insertions(+), 1 deletions(-)

diff --git a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/DistributedFileSystem.java b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/DistributedFileSystem.java
index 3541488..23e158d 100644
--- a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/DistributedFileSystem.java
+++ b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/DistributedFileSystem.java
@@ -975,6 +975,9 @@ public class DistributedFileSystem extends FileSystem {
    * @return return true if file is closed
    * @throws FileNotFoundException if the file does not exist.
    * @throws IOException If an I/O error occurred     
+   * @throws UnsupportedOperationException if the HDFS cluster this client is
+   *          connecting to is running an older version which does not support
+   *          this feature.
    */
   public boolean isFileClosed(Path src) throws IOException {
     return dfs.isFileClosed(getPathName(src));
diff --git a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/protocolPB/ClientNamenodeProtocolTranslatorPB.java b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/protocolPB/ClientNamenodeProtocolTranslatorPB.java
index 200f9f7..506b796 100644
--- a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/protocolPB/ClientNamenodeProtocolTranslatorPB.java
+++ b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/protocolPB/ClientNamenodeProtocolTranslatorPB.java
@@ -853,7 +853,13 @@ public class ClientNamenodeProtocolTranslatorPB implements
     try {
       return rpcProxy.isFileClosed(null, req).getResult();
     } catch (ServiceException e) {
-      throw ProtobufHelper.getRemoteException(e);
+      IOException ioe = ProtobufHelper.getRemoteException(e);
+      if (ioe.getMessage().startsWith("Unknown method isFileClosed called")) {
+        throw new UnsupportedOperationException(
+            "Remote server does not implement isFileClosed.", ioe);
+      } else {
+        throw ioe;
+      }
     }
   }
 
-- 
1.7.0.4

